test-src/org/pentaho/reporting/platform/plugin/cache/FileSystemCacheBackendTest.java

```diff
@@ -135,7 +148,7 @@ public String getValue() {
 
     private void writeObject( final ObjectOutputStream oos ) throws IOException {
       try {
-        Thread.sleep( 3000L );
+        Thread.sleep( 1000L );
```

Comment:
That is what locks are for. Use two locks/latches for enforcing the right order. 

For instance for "testWriteWhileReading" you can enforce the order by setting and resetting latches while you are in readObject(..). These latches must be static, or you could not get a reference to them, but otherwise it is just standard code:

```Java

  @Test
  public void testWriteWhileReading() throws Exception {
    final String val = "ShinySecretValue";
    final String val2 = "ShinySecretValue2";
    final ExecutorService executor = Executors.newFixedThreadPool( 2 );
    readStartedSignal = null;
    writeStartedSignal = null;
    // setup writes without any signals ...
    fileSystemCacheBackend.write( Arrays.asList( directoryKey, key ), new LongRead( val ) , new HashMap<String, Serializable>());
    // now we can setup the signals ..
    readStartedSignal = new CountDownLatch(1);
    writeStartedSignal = new CountDownLatch(1);

    final Future<LongRead> read = executor.submit( new Callable<LongRead>() {
      @Override public LongRead call() throws Exception {
        return (LongRead) fileSystemCacheBackend.read( Arrays.asList( directoryKey, key ) );
      }
    } );
    final Future<Boolean> write = executor.submit( new Callable<Boolean>() {
      @Override public Boolean call() throws Exception {
        readStartedSignal.await();
        // unblock the reader ...
        writeStartedSignal.countDown();
        return fileSystemCacheBackend.write( Arrays.asList( directoryKey, key ), new LongRead( val2 ), new HashMap<String, Serializable>() );
      }
    } );
    assertTrue( write.get() );
    assertEquals( read.get().getValue(), val );
  }

  @After
  public void clearLatchesAfterEachTest() {
    readStartedSignal = null;
    writeStartedSignal = null;
  }

  private static CountDownLatch readStartedSignal;
  private static CountDownLatch writeStartedSignal;


  private class LongRead implements Serializable {

    private String value;

    public LongRead( final String value ) {
      this.value = value;
    }

    public String getValue() {
      return value;
    }

    private void writeObject( final ObjectOutputStream oos ) throws IOException {
      oos.writeObject( value );
    }

    private void readObject( final ObjectInputStream ois ) throws ClassNotFoundException, IOException {
      try {
        if ( readStartedSignal != null ) {
          // unlock the writer ...
          readStartedSignal.countDown();
        }
        if ( writeStartedSignal != null ) {
          // wait for the writer to start working ..
          writeStartedSignal.await();
        }
        // now that small time frame should be enough, as we guarantee that the writer is
        // ready to enter the file-system write method as soon as it sees the latch go.
        Thread.sleep(100);
        final Object o = ois.readObject();
        this.value = (String) o;
      } catch ( final InterruptedException e ) {
        e.printStackTrace();
      }

    }
  }
```